server/src/test/java/org/opensearch/index/engine/CompositeIndexWriterForAppendTests.java

```diff
@@ -267,27 +268,36 @@ public void testConcurrentIndexAndDeleteDuringRefresh() throws IOException, Inte
                         numDeletes.incrementAndGet();
                     }
                 }
-            } catch (Exception e) {
-                throw new AssertionError(e);
+            } catch (Throwable t) {
+                indexerFailure.set(t);
             } finally {
                 done.set(true);
             }
         });
 
-        indexer.start();
-        latch.countDown();
-        latch.await();
-        while (done.get() == false) {
-            compositeIndexWriter.beforeRefresh();
-            compositeIndexWriter.afterRefresh(true);
-            Thread.sleep(100);
-        }
+        try {
+            indexer.start();
+            latch.countDown();
+            latch.await();
+            while (done.get() == false) {
+                compositeIndexWriter.beforeRefresh();
+                compositeIndexWriter.afterRefresh(true);
+                Thread.sleep(100);
+            }
 
-        indexer.join();
-        compositeIndexWriter.beforeRefresh();
-        compositeIndexWriter.afterRefresh(true);
-        try (DirectoryReader directoryReader = DirectoryReader.open(compositeIndexWriter.getAccumulatingIndexWriter())) {
-            assertEquals(numDocs - numDeletes.get(), directoryReader.numDocs());
+            indexer.join();
+            Throwable failure = indexerFailure.get();
+            if (failure != null) {
+                throw new AssertionError("indexer thread failed", failure);
+            }
+            final int expectedDocs = numDocs - numDeletes.get();
+            assertBusy(() -> {
+                compositeIndexWriter.beforeRefresh();
+                compositeIndexWriter.afterRefresh(true);
```

Comment:
Was indexing operation still going on here? Ideally checking this repeatedly should make any difference here. We are flushing and closing the child level IndexWriters before the syncing child writers with parent writers in beforeRefresh. So if no Indexing operation was not in progress post ```indexer.join()```, ideally there should not be any buffered operations.